pkg/bootstrap/installer.go

```diff
@@ -0,0 +1,298 @@
+/*
+Copyright 2023 The Tekton Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package bootstrap
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"os"
+	"os/exec"
+	"time"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+)
+
+type Installer struct {
+	kubeClient kubernetes.Interface
+	config     *Config
+}
+
+func NewInstaller(kubeClient kubernetes.Interface, config *Config) *Installer {
+	return &Installer{
+		kubeClient: kubeClient,
+		config:     config,
+	}
+}
+
+// InstallTriggers installs Tekton Triggers CRDs and controllers
+func (i *Installer) InstallTriggers(ctx context.Context) error {
+	// Check if Triggers is already installed
+	if i.isTriggersInstalled(ctx) {
+		log.Println("Tekton Triggers is already installed.")
+		return nil
+	}
+	if err := i.downloadAndApplyTriggers(ctx); err != nil {
+		return fmt.Errorf("failed to install Triggers: %w", err)
+	}
+
+	log.Println("Waiting for Tekton Triggers to be ready...")
+	return i.waitForTriggersReady(ctx)
+}
+
+func (i *Installer) isTriggersInstalled(ctx context.Context) bool {
+	if i.kubeClient == nil {
+		return false
+	}
+	// Check for the existence of the Triggers namespace and deployments
+	_, err := i.kubeClient.AppsV1().Deployments("tekton-pipelines").Get(ctx, "tekton-triggers-controller", metav1.GetOptions{})
+	return err == nil
+}
+
+// waitForTriggersReady waits for Tekton Triggers components to be ready
+func (i *Installer) waitForTriggersReady(ctx context.Context) error {
+	// Wait for the tekton-pipelines namespace to exist
+	err := wait.PollImmediate(2*time.Second, 5*time.Minute, func() (bool, error) {
+		_, err := i.kubeClient.CoreV1().Namespaces().Get(ctx, "tekton-pipelines", metav1.GetOptions{})
```

Comment:
can namespace be taken dynamically considering `openshift` 🤔 may be check system.Namespace and see what value is coming